File: AtCoder/Waterblues/cpp/ARC107-C-Shuffle_Permutation.cpp
#include <bits/stdc++.h>
#define MAX_N 50
#define MOD 998244353
using namespace std;
typedef long long ll;

ll N, K;
ll A[MAX_N][MAX_N];

ll par[MAX_N], rnk[MAX_N];

void init() {
  for (ll i = 0; i < N; ++i) {
    par[i] = i;
    rnk[i] = 0;
  }
}

ll find(ll x) {
  if (par[x] == x) return x;
  return par[x] = find(par[x]);
}

void unite(ll x, ll y) {
  x = find(x);
  y = find(y);
  if (x == y) return;
  if (rnk[x] < rnk[y])
    par[x] = y;
  else {
    par[y] = x;
    if (rnk[x] == rnk[y]) ++rnk[x];
  }
}

int main() {
  cin >> N >> K;
  for (ll i = 0; i < N; ++i) {
    for (ll j = 0; j < N; ++j) {
      cin >> A[i][j];
    }
  }
  init();
  for (ll i = 0; i < N - 1; ++i) {
    for (ll j = i + 1; j < N; ++j) {
      bool f = true;
      for (ll k = 0; k < N; ++k) {
        if (A[i][k] + A[j][k] > K) f = false;
      }
      if (f) unite(i, j);
    }
  }
  map<ll, ll> C;
  for (ll i = 0; i < N; ++i) {
    ++C[find(i)];
  }
  ll res = 1;
  for (auto p : C) {
    for (ll i = 1; i <= p.second; ++i) {
      res = res * i % MOD;
    }
  }
  init();
  for (ll i = 0; i < N - 1; ++i) {
    for (ll j = i + 1; j < N; ++j) {
      bool f = true;
      for (ll k = 0; k < N; ++k) {
        if (A[k][i] + A[k][j] > K) f = false;
      }
      if (f) unite(i, j);
    }
  }
  C = map<ll, ll>();
  for (ll i = 0; i < N; ++i) {
    ++C[find(i)];
  }
  for (auto p : C) {
    for (ll i = 1; i <= p.second; ++i) {
      res = res * i % MOD;
    }
  }
  printf("%lld\n", res);
}

File: AtCoder/Waterblues/python/ARC107-C-Shuffle_Permutation.py
from collections import Counter
MOD = 998244353
N, K = map(int, input().split())
A = [list(map(int, input().split())) for _ in range(N)]

par, rnk = [], []


def init():
    global par, rnk
    par = list(range(N))
    rnk = [0] * N


def find(x):
    global par
    if par[x] == x:
        return x
    par[x] = find(par[x])
    return par[x]


def unite(x, y):
    x = find(x)
    y = find(y)
    if x == y:
        return
    if rnk[x] < rnk[y]:
        par[x] = y
    else:
        par[y] = x
        if rnk[x] == rnk[y]:
            rnk[x] += 1


init()
for i in range(N - 1):
    for j in range(1, N):
        f = True
        for k in range(N):
            if A[i][k] + A[j][k] > K:
                f = False
        if f:
            unite(i, j)
C = Counter()
for i in range(N):
    C[find(i)] += 1
res = 1
for v in C.values():
    for i in range(1, v + 1):
        res = res * i % MOD
init()
for i in range(N - 1):
    for j in range(1, N):
        f = True
        for k in range(N):
            if A[k][i] + A[k][j] > K:
                f = False
        if f:
            unite(i, j)
C = Counter()
for i in range(N):
    C[find(i)] += 1
for v in C.values():
    for i in range(1, v + 1):
        res = res * i % MOD
print(res)
